### PR TITLE
Support replacing files in `rename` for better compatibility with POSIX

### PR DESCRIPTION
**Pull request by s5bug**

### CODE REVIEW

File: libctru/source/archive_dev.c [hunk 1]
```diff
@@ -980,12 +980,23 @@ archive_rename(struct _reent *r,
   }
 
   rc = FSUSER_RenameFile(sourceDevice->archive, fs_path_old, sourceDevice->archive, fs_path_new);
-  if(R_SUCCEEDED(rc))
-    return 0;
+  /* if the file at the target destination exists, overwrite it */
+  if(rc == 0xC82044BE) {
```
Comment (WinterMute) [inline libctru/source/archive_dev.c:984]: I don't like the magic numbers here. We have some macros that might be useful in https://github.com/devkitPro/libctru/blob/master/libctru/include/3ds/result.h

I think `if ( R_FAILED(rc) && R_DESCRIPTION(rc) == RD_ALREADY_EXISTS`) should probably do it.

File: libctru/source/archive_dev.c [hunk 2]
```diff
@@ -980,12 +980,23 @@ archive_rename(struct _reent *r,
...
+    if(R_FAILED(rc)) return -1;
+    rc = FSUSER_RenameFile(sourceDevice->archive, fs_path_old, sourceDevice->archive, fs_path_new);
+    if(R_SUCCEEDED(rc)) return 0;
+  } else if(R_SUCCEEDED(rc)) return 0;
 
   rc = FSUSER_RenameDirectory(sourceDevice->archive, fs_path_old, sourceDevice->archive, fs_path_new);
-  if(R_SUCCEEDED(rc))
-    return 0;
+  /* if the directory at the target destination exists, overwrite it */
+  if(rc == 0xC82044BE) {
```
Comment (WinterMute) [inline libctru/source/archive_dev.c:993]: Same with this one.

**Review by WinterMute — CHANGES_REQUESTED** [review]

File: libctru/source/archive_dev.c [hunk 3]
```diff
@@ -980,12 +980,23 @@ archive_rename(struct _reent *r,
   }
 
   rc = FSUSER_RenameFile(sourceDevice->archive, fs_path_old, sourceDevice->archive, fs_path_new);
-  if(R_SUCCEEDED(rc))
-    return 0;
+  /* if the file at the target destination exists, overwrite it */
+  if(rc == 0xC82044BE) {
+    rc = FSUSER_DeleteFile(sourceDevice->archive, fs_path_new);
+    if(R_FAILED(rc)) return -1;
```
Comment (piepie62) [inline libctru/source/archive_dev.c:986]: Errno should be set here

File: libctru/source/archive_dev.c [hunk 4]
```diff
@@ -980,12 +980,23 @@ archive_rename(struct _reent *r,
...
+  } else if(R_SUCCEEDED(rc)) return 0;
 
   rc = FSUSER_RenameDirectory(sourceDevice->archive, fs_path_old, sourceDevice->archive, fs_path_new);
-  if(R_SUCCEEDED(rc))
-    return 0;
+  /* if the directory at the target destination exists, overwrite it */
+  if(rc == 0xC82044BE) {
+    /* only overwrite empty directories */
+    rc = FSUSER_DeleteDirectory(sourceDevice-> archive, fs_path_new);
+    if(R_FAILED(rc)) return -1;
```
Comment (piepie62) [inline libctru/source/archive_dev.c:996]: Same here

**Review by piepie62 — CHANGES_REQUESTED** [review]

**Review by WinterMute — APPROVED** [review]

**Review by fincs — APPROVED** [review]

**Pull request merged**